wallet/datastore.go

```diff
@@ -559,20 +559,32 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
 
 // LinkWallet links a wallet together
 func (pg *Postgres) LinkWallet(ctx context.Context, ID string, userDepositDestination string, providerLinkingID uuid.UUID, anonymousAddress *uuid.UUID, depositProvider string) error {
-	sublogger := logger(ctx).With().
-		Str("wallet_id", ID).
-		Logger()
+	sublogger := logger(ctx).With().Str("wallet_id", ID).Logger()
 
 	// create tx
 	tx, err := createTx(ctx, pg)
 	if err != nil || tx == nil {
-		sublogger.Error().Err(err).
-			Msg("error creating tx for wallet linking")
+		sublogger.Error().Err(err).Msg("error creating tx for wallet linking")
 		return fmt.Errorf("failed to create tx for wallet linking: %w", err)
 	}
+
+	err = waitAndLock(ctx, tx, providerLinkingID)
+	if err != nil {
+		sublogger.Error().Err(err).Msg("link wallet error acquire lock")
+		return fmt.Errorf("failed to acquire lock link wallet : %w", err)
+	}
+
 	// add tx to ctx for future
 	ctx = context.WithValue(ctx, appctx.DatabaseTransactionCTXKey, tx)
-	defer pg.RollbackTx(tx)
+	defer func() {
+		if err != nil {
+			err = unlock(ctx, tx, providerLinkingID)
```

Comment:
will a rollback in the Tx cause this lock to disappear?

Reply:
yes, advisory locks dont follow tx but we call unlock to handles this they are also auto cleaned up by pg once session disconnects